flax/linen/module.py

```diff
@@ -396,6 +396,54 @@ def _get_unbound_fn(method_or_fn: Callable[..., Any]) -> Callable[..., Any]:
 
   return method_or_fn
 
+def _validate_field_value(
+  name: str, val: str, module: "Module", variables: Mapping[str, Mapping[str, Any]]
+):
+  """Determines if a value created by `.param()` or `.variable()`
+  was assigned to a field with the same name during `setup`. This logic
+  will cause the following code to raise an error:
+
+  Example::
+
+    def setup(self)
+      self.baz = self.param('bar', ...) # Error
+
+  """
+  # check for Module names
+  modules = (
+    m for m in jax.tree_leaves(val, is_leaf=lambda x: isinstance(x, Module)) 
+    if isinstance(m, Module)
+  )
+
+  for module in modules:
+    if module.name is not None and module.name != name:
+      raise RuntimeError(
+        f'Module name \'{module.name}\' does not match field name \'{name}\''
+      )
+
+  
+  value_found = False
```

Comment:
I'm not sure how this deals with nested params, e.g.: self.theta = {"kernel": self.param("kernel", ...)}

Reply:
#2100 is trying to solve very simple breaks in correspondance:

```python
class Foo(nn.Module):
  def setup(self):
    self.foo = nn.Dense(1, name="bar")
    self.qup = self.param('baz', lambda k: jnp.zeros((1,)))
```
Your example involves additional operations over the value which I would consider valid. Base case for this would be:

```python
self.bar = self.param("bar", lambda key: jnp.array(1))
self.baz = self.bar + 1 # valid 
self.foo = {'foo': self.baz} # valid
self.fii = self.foo # recently made valid
```